Plans 2023: Decouple CTA actions from plans-grid-next package ( Attempt 2 )

jeyip opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/pull/86886

## Proposed Changes

Description taken from https://github.com/Automattic/wp-calypso/pull/86886 and amended where appropriate for this second attempt.

This PR is the first step to fully decouple the CTA configuration of the plans grid, inspired by @jdc91 's post pdvytD-qk-p2. Conceptually, it's an extension to the current `PlanActionOverrides` system, but it only replaces the callback functions part for now as the 1st step. 
The ultimate goal is that the `plans-grid-next` package simply offers the `PlanActions` prop for the calling site to configure the clicked actions, text, disabled or not, etc. The package itself shouldn't need to know those contextual info.

Summarizing the introduced changes:

* Create a `useActionCallback` hook that encapsulates all logic that determines what CTA function to use in a given circumstance
* Note that the upgrade handler can obviously be simplified, but the current structure is generally maintained for now so as to avoid introducing too much change at once.
* Introduce an action handler mechanism, i.e. `showModalAndExit`, which allows the calling site to introduce further actions on the CTA click. It also allows returning a boolean value to indicate whether it wants to trigger early return, so the actual shopping cart processing won't happen. Two cases supported by this are the domain/plan upsell modal and the domain upsell from home modal.
* Deprecate the current CTA callback props, e.g. onUpgradeClick, handleUpgradeClick, `callback` in `PlanActionOverrides`, etc.
* The actual on-click callback has been separated finer, e.g. managing add-ons, managing plans, redirecting to the VIP page, handling upgrades, instead of being combined into one like it currently does. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Onboarding Ctas**
- Navigate to `/start/plans`
- Smoke test plan ctas in the features grid ( free, explorer, creator, etc. ) and ensure that the correct plan is added to the checkout cart
- Smoke test creator or entrepreneur plan ctas with a storage add-on and ensure the correct products are added to the checkout cart
- Repeat for the comparison grid
- Lists of flows with the plans pricing page can be found here PCYsg-RDT-p2
- Repeat for a different signup flow ( Ex. `start/domain` ) 
- Repeat for a stepper flow ( Ex. `setup/newsletter` ) 

**Paid Domain Modal Upsell Ctas**
- Navigate to `/start/onboarding-pm`
- Select a paid, custom domain
- Select a free plan
- Verify that the upsell modal is displayed as expected
- Smoke test the ctas in the upsell modal

<img width="725" alt="Screenshot 2024-04-23 at 4 04 39 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/537ca848-ac0a-4c99-b1fc-d5c4b08dc59c">

**Free Domain Free Plan Modal Upsell Ctas**
- Navigate to `/start/onboarding-pm`
- Select a free domain
- Select a free plan
- Verify that the upsell modal is displayed as expected
- Smoke test the ctas in the upsell modal

<img width="664" alt="Screenshot 2024-04-23 at 4 04 10 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/dc8af121-2e1a-401d-85f6-0e72701faa79">

**Plans Page Ctas**
- Navigate to `/plans` for a specific site Ex. `/plans/testsite123.wordpress.com`
- Smoke test plan ctas ( free, explorer, creator, etc. ) and ensure that the correct plan is added to the checkout cart
- Smoke test creator or entrepreneur plan ctas with a storage add-on and ensure the correct products are added to the checkout cart
- Repeat for the comparison grid
- Smoke test ctas for the spotlight plan ( plan card rendered at the top of the page )
  - Ensure the manage add-ons button redirects to the `/add-ons` page
  - Ensure the manage plan button redirects to the plan management page

**Test Free Trials Plan Ctas**
- Navigate to `/setup/new-hosted-site`
- Verify that the Creator plan free trial CTA behaves as it does in production ( leads to the `/trialAcknowledge` page )
- Verify that a free trial site is created when continuing through onboarding
- Smoke test the other Creator and Entrepreneur plan CTA buttons

<img width="824" alt="Screenshot 2024-04-23 at 4 12 31 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/560b9521-3fb4-4e2d-88f5-a353290df9e8">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

jeyip committed:
Introduce `use-plan-actions` hook

Remove `use-upgrade-click-handler` hook

Fix onPlanCtaClick CTA

Create `usePlanAction` hook

Rename `use-plan-action` hook to `use-action-callback`

jeyip committed:
Remove planActions hook

jeyip committed:
Fix getActionCallback interface and bugs

Add missing useMemo dependencies

Remove `PlanActions` type from comparison grid

Add `storageAddOn` param to `onPlanCtaClick`

Fix getActionCallback type error

Add options to getActionCallback

Remove unnecessary `PlanAction` type

Add `useSelectedStorageAddOn` hook to `plans-grid-next`

Consolidate params for `getActionCallback`

Rename `addSelectedPlanAndStorageAddOn` to `processCartItemsForPlansAndAddOns`

chriskmnds commented on packages/plans-grid-next/src/types.ts:
```diff
@@ -138,35 +134,31 @@ export interface ComparisonGridProps extends CommonGridProps {
 	intent?: PlansIntent;
 	selectedSiteId?: number | null;
 	useCheckPlanAvailabilityForPurchase: Plans.UseCheckPlanAvailabilityForPurchase;
+	getActionCallback: ( planSlug: PlanSlug, options?: ActionCallbackOptions ) => () => void;
```
What would it take to turn this into a hook -> `useActionCallback`?

jeyip replied:
Good question -- there are several params in available in Plans Features Main needed to generate the different kind of action callbacks which can be invoked by a CTA. These params _aren't_ currently passed into the plans grids ( ex. `withDiscount`, `flowName`, `sitePlanSlug`, etc. ). 

What's happening now is that those params are passed into the `useActionCallback` hook, at which point they exist in a closure for the subsequent `getActionCallback` function returned by said hook. This function, along with its closure, is then passed into the plans grid.

I'll think this over, but my initial impression was that we would need to inject all of those extra params into the Plans Grid to keep things as a hook

https://github.com/Automattic/wp-calypso/pull/88694/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR388-R396

chriskmnds replied:
> My thought was to inject those params into the useActionCallback hook, at which point they would exist in a closure for the returned getActionCallback function so we wouldn't have to pass them all into the plans grid.

Yes, that sounds like the right way. I was just wondering if we could pass a hook somehow, obviously without passing all these props to the grids. Like the `useCheckPlanAvailabilityForPurchase` hook above 🤔 

If it makes sense... it might just be a case of renaming the consumer end hook. Let's maybe come back to the idea when everything's ready :-)

jeyip replied:
_Edit:_
Apologies! For some reason your reply above ( https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1545483629 ) didn't load so I didn't read it when I drafted the comment below. 😅 I'm taking a closer look now.

_Edit 2:_
I'll take a look at `useCheckPlanAvailabilityForPurchase` soon and yes, we can have a deeper discussion when the PR is ready for review :) I see exactly what you're saying now

---

So after looking more deeply into this, I currently see maintaining the hook in one of two ways:

### Injecting New Props into the Plans Grid
Continuing the thoughts from the comment above https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1541600521

We can pass [several new props]( https://github.com/Automattic/wp-calypso/pull/88694/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR389-R395 ) into the plans grid ( `withDiscount`, `flowName`, `sitePlanSlug`, etc. ) that are used to [generate the different kinds of CTA functions]( https://github.com/Automattic/wp-calypso/pull/88694/files#diff-414f2908c4fdd943075d97bbf756af6a4cc2a5b0faf5fdd045108cb0cb4b4974R134-R175 ) we invoke at runtime. We would then pass these params as additional arguments when invoking `useActionCallback`. Not the biggest fan of this approach because it feels like a roundabout way of generating the eventual callback CTA, however I'm open to thoughts on this.

There are variations on this approach, like generating the different possible CTA functions and passing a dictionary of them as props into the plans grid instead. We would then pass the dictionary as arguments into `useActionCallback`

### Creating a Callback that returns a `useActionCallback` hook
As mentioned in https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1541600521, we currently create a closure of params for the `getActionCallback` function returned by the hook `useActionCallback` hook. 

We could, instead, reverse the approach, and return a `useActionCallback` hook from a `getUseActionCallbackHook` function ( where the function would create a closure for the hook instead of vice versa).

 Either way, I do see the value in passing a hook into the plans grid instead of a function, and would love your eventual thoughts on this when this PR is ready for review.

chriskmnds replied:
Thanks for looking into this!

> We could, instead, reverse the approach, and return a `useActionCallback` hook from a `getUseActionCallbackHook` function ( where the function would create a closure for the hook instead of vice versa).

This sounds like a good approach. So we can stick to hook form on callbacks passed into the grid.

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,195 @@
+		[ siteSlug, withDiscount, planActionCallback, cartHandler ]
+	);
+
+	const addSelectedPlanAndStorageAddon = useCallback(
```
Could this be `processCartItemsForPlan` or something of this form? just to give some more immediate feedback on what it does internally?

jeyip replied:
Yes I'll take a closer look at naming 👍

jeyip replied:
Renamed to `processCartItemsForPlanAndAddOns`. It's still as verbose as the last name, but I think it's much clearer what the function is doing now. Thanks for the recommendation

jeyip replied:
[5d99659](https://github.com/Automattic/wp-calypso/pull/88694/commits/5d996598ada0645ae9a14109edf8edb8bf4f99a0)

chriskmnds commented on packages/plans-grid-next/src/types.ts:
```diff
@@ -138,35 +134,31 @@ export interface ComparisonGridProps extends CommonGridProps {
 	intent?: PlansIntent;
 	selectedSiteId?: number | null;
 	useCheckPlanAvailabilityForPurchase: Plans.UseCheckPlanAvailabilityForPurchase;
+	getActionCallback: ( planSlug: PlanSlug, options?: ActionCallbackOptions ) => () => void;
```
Instead of two arguments, can we turn this into one? Thinking of something like

```
interface UseActionCallbackParams {
    planSlug: PlanSlug;
    cartItemForPlan?: { product_slug: string } | null;
    currentPlan?: boolean;
    freeTrialPlanSlug?: PlanSlug;
    isFreeTrialPlan?: boolean;
    selectedStorageAddOn?: AddOns.AddOnMeta | null;
}
```

jeyip replied:
We totally can 👍 Happy to do this

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/d3c9d5507ff1d0071286a939605b346c8d4152e7

chriskmnds commented on packages/plans-grid-next/src/types.ts:
```diff
@@ -138,35 +134,31 @@ export interface ComparisonGridProps extends CommonGridProps {
 
+export type ActionCallbackOptions = {
+	cartItemForPlan?: { product_slug: string } | null;
+	currentPlan?: boolean;
```
This sounds like something we can derive at the consuming end, so no need to extend the interface here to accept it as a param. WDYT? 🤔

jeyip replied:
Pausing responses to [take a look at another priority ]( p1711551683801899/1711011081.834139-slack-C06FCT6EEHH )-- will return to this when I have a bit more time

jeyip replied:
Didn't get too far into these changes today, but thanks for the patience -- I think I'll have a PR up for more thorough review by eod tomorrow

jeyip replied:
> This sounds like something we can derive at the consuming end, so no need to extend the interface here to accept it as a param. WDYT? 🤔

Ah so if I understand correctly, our goal is to simplify the interface more 🤔 I _do_ wholeheartedly agree with this idea. Hoping to confirm what this looks like in practice, because I think I'm missing a step here.

~`currentPlan` identifies a spotlight plan, and dictates what kind of function is invoked for the CTA ( manageAddOn / managePlan or upgrade cart items  ) see [here]( https://github.com/Automattic/wp-calypso/pull/88694/files#diff-414f2908c4fdd943075d97bbf756af6a4cc2a5b0faf5fdd045108cb0cb4b4974R193-R195 )~

~From what i see about deriving at the consuming end, doing so seems like it would muddy responsibilities of our `getActionCallback` function ( logic for determining which CTA to invoke would now exist in the consuming end _and_ the executing end ) Curious if you could elaborate a bit more~

_Edit:_

I think I see where my confusion was at. I believe when you say consuming end, you're talking about usage, or consumption, of the params within the `useActionCallback` hook and `getActionCallback` function. You're saying we would derive the values with store selectors or through other means, which I agree with 👍 

I was confusing "consuming end" with the concept of the _consumer of a function_, which was throwing me for a loop.

I'll make the changes :D

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/f881e55d4a1f2f64b7c0553fad43f12bee0c5337

chriskmnds replied:
Hey @jeyip I left a couple of comments on the commit that you linked to concerning `currentPlan`. I don't see them - let me know if they came through or if I should repost on the main changes instead. :D

jeyip replied:
Hey thanks for the heads up. I saw them _and_ responded to them, and then subsequently lost them again 😆

Finally found them for a second time, so here's a link for posterity https://github.com/Automattic/wp-calypso/commit/f881e55d4a1f2f64b7c0553fad43f12bee0c5337#r140558754

chriskmnds commented on packages/plans-grid-next/src/types.ts:
```diff
@@ -138,35 +134,31 @@ export interface ComparisonGridProps extends CommonGridProps {
+export type ActionCallbackOptions = {
+	cartItemForPlan?: { product_slug: string } | null;
+	currentPlan?: boolean;
+	freeTrialPlanSlug?: PlanSlug;
+	isFreeTrialPlan?: boolean;
```
Same as above? Do you think deriving these from the consuming end and keeping the signature smaller might be better (or worse)?

jeyip replied:
Agree -- I'll take a look :)

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/7b5b0af341de89f8d4b59f124f3910bca946e981

chriskmnds reviewed (commented):
Thanks! It certainly looks like heading in the right direction to me :-)

I left some high-level comments, mainly looking at the types/props and reading some of the code. The consuming end hook looks a bit complex at the surface, but we can get to that later / when ready for review? Happy to study the code closer when you like.

How are we planning to extract/deprecate `PlanActionOverrides`. Are the two going to be aligned into a single prop?

jeyip commented:
Apologies for the borked rebase y'all! Fixing things now

jeyip committed:
Implement useActionCallback hook

Derive `freeTrialPlanSlug` in the useActionCallback hook

Derive storage add on feature slug in use action callback hook

Fix paid media free plan CTA

Fix actions component tests

Remove planActionOverrides loggedInFreePlan callback

Remove currentPlan planActionOverride callback

Implement `Plans.useCurrentPlan` hook in `useActionCallback`

Replace `getActionCallback` function with `useActionCallback` hook

Remove named object props from UseActionCallback

Move `calypso_signup_free_plan_click` tracks event into upgrade handler callback

Remove `freeTrialCartItem` temp variable

Add `return` to signal end of cart item processing

Update `planActionCallback` return value type

Update `useActionCallback` to `useGenerateActionCallback`

Remove unnecessary freeTrialPlanSlug check

Update manageAddon to manageAddOn

Fix types

Import `recordTracksEvent` from `@automattic/calypso-analytics` package

Remove `useActionCallbacks` hook

Destructure useActionCallbackParams in useActionCallback

Move `useFreeTrialPlanSlugs` into `useActionCallback`

Include VIP UTM params in gotoVip redirect cta

Remove `composeActionCallback` helper

Remove `processCartItemsForPlanAndAddOns` helper function

jeyip commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,203 @@
+import page from '@automattic/calypso-router';
+import { useMemo, useCallback } from '@wordpress/element';
+import { useSelector } from 'react-redux';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks'; //TODO: move this out
```
@southp I see that you left this comment in some of the logic I copied over from the first attempt at decoupling CTA actions. 

I'm uncertain about your original reasoning though. Could you explain a bit more? To me, `recordTracksEvent` seems fine where it is

chriskmnds replied:
We also use the one from `@automattic/calypso-analytics` and not the local dependency (not essential though, obviously) :-)

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/8c8133ac5c09d025355519c87b0ccf646061b88d

jeyip commented:
#### Note:

The unit tests failing for this PR are not related to our changes and can can be traced back to `trunk` https://github.com/Automattic/wp-calypso/commit/d5c9f50abc3945f9ecebed516a4dd6d5554d13a6. I've reached out to folks to see if they can resolve things :) p1712015076198379-slack-C7YPUHBB2

jeyip commented:
> How are we planning to extract/deprecate PlanActionOverrides. Are the two going to be aligned into a single prop?

Now that CTA callback logic is _finally_ in a good spot for review, I have time to circle back to this important question. Thanks for the patience 🙂

Yes I was imagining a single prop similar to what Dehan was originally suggesting pdvytD-qk-p2. This, however, is a good place to discuss alternative approaches as well.

```javascript
// useActions.ts
type Action {
  [ key: string ]: {
    text: string;
    postTextButton?: string;
    status?: 'enabled' | 'blocked';
  }
}

type Actions {
  useActionCallback: () => void;
  loggedInFreePlan: Action;
  currentPlan: Action;
  // More CTA Actions...
}

function useActions( params ): Actions {
  // Logic to generate useActionCallback
  // Logic to generate different actions
  // Each action effectively corresponds to a distinct CTA

  return {
    useActionCallback,
    loggedInFreePlan,
    currentPlan,
    // More actions...
  };
}
  
}

// In the PlansFeaturesMain component
const actions = useActions( params )

<FeaturesGrid actions={actions} />
<ComparisonGrid actions={actions} />
```

Alternatively, we could also keep the concept of the action CTA callback and the action CTA data separated. 

```javascript
// In the PlansFeaturesMain component
const actions = useActionsCtaUI( params )
const useActionCallback = getUseActionCallback( params )

<FeaturesGrid actions={actions} useActionCallback={useActionCallback} />
<ComparisonGrid actions={actions} useActionCallback={useActionCallback} />
```

I'm a little torn on either approach, but I prefer the first because it seems natural to me to keep CTA UI information alongside CTA callback logic in a single data structure. Always open to your suggestions though Christos 🙏

jeyip commented:
Hey @chriskmnds I still need to fix some type errors, but this PR is generally ready for a deeper dive into review. Thanks so much for the guidance and patience so far 🙂

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+				recordTracksEvent( 'calypso_signup_free_plan_click' );
+			}
+
+			const earlyReturn = planActionCallback?.( planSlug );
```
I don't understand this.  why does a truthy value signal an "early return"? it seems rather unconventional.
also type is `() => void`.. 🤔

jeyip replied:
> I don't understand this. why does a truthy value signal an "early return"? it seems rather unconventional.

Thanks for bringing this up -- I think it might be a learning moment for me. I didn't realize that early returns are primarily associated with falsey value signals, but I can see how this might be the case.

I do, however, see examples throughout our codebase where truthy value signals lead to an early function return in [php]( https://github.com/Automattic/wp-calypso/blob/6e7b9d3f90b084c86265e8fe0ec1814fc86c4dc6/apps/editing-toolkit/editing-toolkit-plugin/help-center/class-help-center.php#L41-L43 ), [js]( https://github.com/Automattic/wp-calypso/blob/6e7b9d3f90b084c86265e8fe0ec1814fc86c4dc6/apps/wpcom-block-editor/src/wpcom/features/site-editor-env-consistency.js#L15-L18 ), and [react]( https://github.com/Automattic/wp-calypso/blob/6e7b9d3f90b084c86265e8fe0ec1814fc86c4dc6/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.tsx#L62-L64 ), so I think I might be missing something. Would you mind elaborating on your thoughts a bit more?

> () => void.. 🤔

Taking a look now 👀

jeyip replied:
Thanks for catching the erroneous `void` return value. I updated it in https://github.com/Automattic/wp-calypso/pull/88694/commits/8144e3368a86da6edea82ae52402ea3c88d92eda

jeyip replied:
> I don't understand this. why does a truthy value signal an "early return"?

After reviewing the logic further, it appears that there are situations where we want to short circuit traditional CTA logic ( Ex. when we click on a plan CTA and want to show an upsell dialog instead of adding the plan to the checkout cart )

I see that James attempted to encapsulate all of the messiness into a `planActionCallback` ( chock full of fun side effects 😆  ), probably to reconsider in a future PR. There's a certain amount of indirection that's introduced with this pattern. https://github.com/Automattic/wp-calypso/pull/88694/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR365-R385

My senses tell me that this will be worthy of its own, independent follow-up PR, but this is open for discussion. In the meantime, I'm thinking about an easy way to make things more intuitive.

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+				checkoutUrl
+			);
+
+			page( checkoutUrlWithArgs );
```
Let's follow this with an empty `return;` so that we signal this is the end of processing.

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/785d8bfefddde3619af590f0aae76c223c2e0cc7

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+			const { cartItemForPlan, planSlug, selectedStorageAddOn } = params;
+
+			if ( isFreeTrialCTA && freeTrialPlanSlug ) {
+				const freeTrialCartItem = { product_slug: freeTrialPlanSlug };
```
NIT - I guess we don't really need this variable. just define below in the call?

jeyip replied:
Taking a look 👀

jeyip replied:
Happy to make the change here 🙂 https://github.com/Automattic/wp-calypso/pull/88694/commits/52d0f359e179cdb9851de59ef3acc0588eefd1da

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+			const cartItemForPlan = getPlanCartItem( cartItems );
+			const planSlug = clickedPlanSlug ?? PLAN_FREE;
+
+			if ( isFreePlan( planSlug ) ) {
+				recordTracksEvent( 'calypso_signup_free_plan_click' );
+			}
```
I think this would be better suited below line 127, in an else clause under `if ( ! isFreePlan( planSlug ) ) { ...`?

jeyip replied:
Agreed -- I didn't spend too much time looking at the contents of `useUpgradeHandler` because I figured, similarly to James, that it would be safer to avoid too much change to upgrade CTA logic all at once like he mentions in https://github.com/Automattic/wp-calypso/pull/86886

This change seems obvious though so I'm happy to ahead and make the updates 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/b7ac137cb010a9c0b451a8db1ba72f8546742bd9

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+		) => {
+			const { cartItemForPlan, planSlug, selectedStorageAddOn } = params;
+
+			if ( isFreeTrialCTA && freeTrialPlanSlug ) {
```
Is there any chance we'd have a `isFreeTrialCTA` and not a `freeTrialPlanSlug`? 🤔

jeyip replied:
Ah another good catch. I believe you're right about not needing to check `freeTrialPlanSlug` here, but will double check this first thing tomorrow morning.

jeyip replied:
From what I can tell, no, `isFreeTrialCTA` will not be `true` without a `freeTrialPlanSlug` https://github.com/Automattic/wp-calypso/pull/88694/files#diff-00e19b2171f5efb1c76799b683dad83ab9c2512c3985ec84c20a7d62ad175dd8R120-R123

I'm updating the condition accordingly 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/e49a9eef9956fa298984def964cbc7cb3e6d9408

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,220 @@
+
+				if ( earlyReturn ) {
+					return;
+				}
```
Similar to other comment. How's "early return" defined here?

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1558493941

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -421,6 +363,51 @@ const PlansFeaturesMain = ( {
+		eligibleForFreeHostingTrial,
+	} );
+
+	const actionCallbacks = useActionCallbacks(
+		flowName,
+		sitePlanSlug,
+		siteSlug,
+		withDiscount,
+		planActionCallback,
+		onUpgradeClick
+	);
+	const useActionCallback = getUseActionCallback(
+		actionCallbacks,
+		currentPlan,
+		intent,
+		freeTrialPlanSlugs,
+		sitePlanSlug
+	);
```
I don't understand why these calls have to be done separately and the former fed into the latter. Can't we turn `getUseActionCallback` into a hook? (I'm guessing this is the only reason we are splitting them out?)

It could be `useGenerateActionCallback` or something of that form 🤔

jeyip replied:
Hmmm...lemme revisit this with fresh eyes. 

> Can't we turn getUseActionCallback into a hook? (I'm guessing this is the only reason we are splitting them out?)

You're correct about this. Doing a thorough re-review in a bit.

jeyip replied:
Ah I remember now.

It felt unusual to have a hook ( `useGenerateActionCallback` ) return another hook ( `useActionCallback` ) primarily because it feels like an uncommon pattern that might not be intuitive to folks. 

Because of this, I opted for a function ( `getUseActionCallback` ) that returns a hook ( `useActionCallback` ), but, like you've noted, it required me to create a secondary hook ( `useActionsCallbacks` ) to generate dependent information `actionCallbacks`.

I don't have a strong opinion on my approach, so I'll go ahead and switch to what you've suggested. If we change our minds, it should be easy to rework.

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/781c3c51d0e2ef500b5a6b3a0de8870c0d70f02

chriskmnds commented on packages/plans-grid-next/src/types.ts:
```diff
@@ -138,37 +133,32 @@ export interface ComparisonGridProps extends CommonGridProps {
+	selectedStorageAddOn?: AddOns.AddOnMeta | null;
+};
+
+export type UseActionCallback = ( params: UseActionCallbackParams ) => () => void;
```
Curious - I'd have expected something like the following:

```
export type UseActionCallback = ( { planSlug, cartItemForPlan, selectedStorageAddOn }: UseActionCallbackParams ) => () => void;
```

so to not use a named object for the props

jeyip replied:
Oh this didn't even cross my mind. I completely agree with you here. I think I had been too close to the code for too long and this followed a string of other changes 🤔 

Lemme make the update 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/f2f1bbc98dfa1f890fbff24961058982d3495d40

chriskmnds reviewed (commented):
Hey! I left a few more comments, mostly nitpicks. I still find the `use-action-callback` part a little hard to follow, so need to revisit with a more clear mind. Probably a little later/tomorrow :-)

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -421,6 +362,47 @@ const PlansFeaturesMain = ( {
+		}
+
+		return false;
+	};
```
I think at least part of the complexity and indirection results from this living here and being passed into the other hook. It should be possible to either migrate it entirely out by adding `isModalOpen` state to the wpcom-plans-ui store. Otherwise, I'd consider renaming it to something more specific, we may be overdoing it a bit with the generalised "callback" term - almost every new function has the name "callback" in it and nothing else (or another generalised term, like "compose").

jeyip replied:
I've been mulling this over and still don't have a great answer for the naming 🤔 

It's a bit challenging because the function is serving two very distinct responsibilities. It short circuits normal cta logic _and_ it also shows some modal or upsell.

Perhaps `exitEarlyAndShowModal`? Thinking this over a bit more.

chriskmnds replied:
Yeah, naming this will be hard. Something tells me we need to reorganize things a little (wrap differently) and avoid passing it through as a prop to be called from the other hook. We don't need to "centralize" everything.

I can revisit/try something tomorrow if nothing else comes to mind.

jeyip replied:
Hey Christos unfortunately I didn't get to this today, but will start focusing on this plan action callback first thing tomorrow

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,234 @@
+		withDiscount,
+		planActionCallback,
+		cartHandler
+	);
```
Is this abstraction necessary? I feel it results in additional complexity, while not adding/saving anything. The named callbacks returned feel like a complication for the sake of having another function (with additional costs and memoization).

jeyip replied:
Good question 👍 I don't have strong opinions about my approach here. 

A bit more about my reasoning -- I tend to prefer grouping logic behind declarative helper functions, especially if they're in the same file. Admittedly, `useActionCallbacks` isn't very descriptive so it could add to confusion rather than alleviate it. I also acknowledge that the approach adds a certain amount of indirection, like having to hop between different parts of the page to find the right function, especially if naming doesn't properly clarify what the function is doing.

With that being said, concerns about memoization, current readability, and overusage of the term callback are valid, and I'm more than happy to remove the abstraction 😎

jeyip replied:
[a87a1dd](https://github.com/Automattic/wp-calypso/pull/88694/commits/a87a1dd7916e38810a1ec4eadc9eb0b60cce9681)

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,234 @@
+	);
+}
+
+function composeActionCallback(
+	callback: () => void,
+	planActionCallback?: ( planSlug: PlanSlug ) => boolean
+) {
+	return ( { planSlug }: Pick< UseActionCallbackParams, 'planSlug' > ) => {
+		const earlyReturn = planActionCallback?.( planSlug );
+		return earlyReturn ? () => {} : callback;
+	};
+}
```
Likewise. I am not sure I see the need for this extra function, especially under this name. It simply checks if some other callback returns true/false and sends back the callback passed to it. 

I think we are also overdoing it on abstractions around the name "..callback". It's confusing to me :/

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1559917358

jeyip replied:
For this specific case, there is [a noticeable amount of duplication]( https://github.com/Automattic/wp-calypso/pull/88694/files#diff-414f2908c4fdd943075d97bbf756af6a4cc2a5b0faf5fdd045108cb0cb4b4974R171-R173 ) without the function. The abstraction for this specific case makes sense to me. For the time being, I just [moved it closer to where it's used]( https://github.com/Automattic/wp-calypso/pull/88694/commits/a87a1dd7916e38810a1ec4eadc9eb0b60cce9681d ) for readability, but we can definitely discuss things further.

Giving this more thought, one other point of confusion I _think_ I see with the term `callback` is that, for example, `composeActionCallback` can be read as _either_ "compose action-callback" or "compose-action callback". Perhaps if we used the term Cta instead Ex. `composeActionCta`?

chriskmnds reviewed (commented):
Hey! I left a few more comments. Apologies if being a bit abrupt/direct, it's not my intention. just wrote out my impressions. I can take more/closer looks as we discuss/entertain concerns :-)

No need to change anything if you don't completely agree. I'll check back tomorrow.

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -421,6 +364,47 @@ const PlansFeaturesMain = ( {
+	const freeTrialPlanSlugs = useFreeTrialPlanSlugs( {
+		intent: intent ?? 'default',
+		eligibleForFreeHostingTrial,
+	} );
```
You can call this from `useGenerateActionCallback` right? So no need to pass as a prop.

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/f067f3da3a382f6317990f7dfe93915812d33ced

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,204 @@
+	}, [ currentPlanManageHref, flowName, siteSlug ] );
+
+	return (
+		params: UseActionCallbackParams = {
```
I would use an object spread instead of named `params` prop.

jeyip replied:
Made the changes 👍 https://github.com/Automattic/wp-calypso/pull/88694/commits/4da64707d240be63bddac742a734dd4e2326d764

If you have time, would love to learn a bit more about your reasoning. Otherwise, no worries, I totally understand how busy you probably are. 🙂

chriskmnds replied:
for this case - I think primarily consistency. There's no need to have this other "named thing" in the code.

For future, if you'd like my opinion -> always go with an object spread `{ param, param1, param2, ... }`. You don't need to worry about order, `undefined` props, easier to extend (add another prop) and maintain, more consistent with React, etc.

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,204 @@
+			selectedStorageAddOn: undefined,
+		}
+	) => {
+		const { planSlug } = params;
+		const planConstantObj = applyTestFiltersToPlansList( planSlug, undefined );
+		const freeTrialPlanSlug = freeTrialPlanSlugs?.[ planConstantObj.type ];
+
+		if ( planSlug && isWpcomEnterpriseGridPlan( planSlug ) ) {
+			return gotoVip( params );
+		}
+
+		if ( sitePlanSlug && currentPlan?.productSlug === planSlug && intent !== 'plans-p2' ) {
+			return isFreePlan( sitePlanSlug ) ? manageAddOn( params ) : managePlan( params );
+		}
+
+		return upgradeHandler( params, freeTrialPlanSlug );
```
I wonder if handling this a little differently might help avoid `composeActionCallback` and all the named callbacks. Abstractions aren't always beneficial, and I feel in this case there's more complexity being introduced. More so if something changes e.g. no longer sending to "vip"..

Something along these lines:

```
...
const earlyReturn = planActionCallback?.( planSlug );

if ( earlyReturn ) return;

/* 1. Send user to VIP if it's an enterprise plan */
if ( isWpcomEnterpriseGridPlan( planSlug ) ) {
    recordTracksEvent( 'calypso_plan_step_enterprise_click', { flow: flowName } );
    window.open( 'https://wpvip.com/wordpress-vip-agile-content-platform', '_blank' );
    return;
} 

/* 2. Send user to either manage add-ons or plan in case of current plan selection */
if ( currentPlan?.productSlug === planSlug && intent !== 'plans-p2' ) {
    if ( isFreePlan( planSlug ) {
        page.redirect( `/add-ons/${ siteSlug }` );
    } else {
        page.redirect( currentPlanManageHref );
    }
    return;
}

/* 3. Handle a plan upgrade */
handleUpgrade( params, freeTrialPlanSlug );
return;

```

1. No need to call `planActionCallback?.( planSlug )` from `processCartItems` above either. Just here should suffice from the looks of it. or is that not the case? 🤔 
2. renamed `upgradeHandler` to `handleUpgrade`, to be more in line with "some action happening".

jeyip replied:
@chriskmnds After seeing your writeup, I agree. It's noticeably more readable. 👍 

> Abstractions aren't always beneficial, and I feel in this case there's more complexity being introduced

I 100% agree with your statement about abstractions, and I'm _very_ curious about the how you intuited that there was complexity. Was it your "spidey sense" again that nudged you in this direction? Or maybe it was simply difficult to keep track of the train of logic?

I'm hoping to understand so that I can more clearly recognize it and address it in the future 🙏 

> No need to call planActionCallback?.( planSlug ) from processCartItems above either. Just here should suffice from the looks of it. or is that not the case? 🤔

I believe it will cover that case as well, yes. I'm double checking shortly.

jeyip replied:
> No need to call planActionCallback?.( planSlug ) from processCartItems above either. Just here should suffice from the looks of it. or is that not the case? 🤔

It's requiring a modest amount of refactoring because the `planSlug` in `planActionCallback?.( planSlug )` might be a traditional `planSlug` or a `freeTrialPlanSlug` depending on what CTA is clicked at runtime. This _does_ affect the result of the early return. This would mean that we would have to move early return logic back into the `handleUpgrade` function.

I'm making some updates now to accommodate our changes described in https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1561079787, and you can take a look at them tomorrow to see if they make sense

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/2a04d27cc3c0248406a7c275357cc91aab76f18e and https://github.com/Automattic/wp-calypso/pull/88694/commits/9377f1483b037cb2811a389bbd4b9a149efcd9fe

chriskmnds replied:
> Was it your "spidey sense" again that nudged you in this direction? Or maybe it was simply difficult to keep track of the train of logic?

Not sure. It seemed like something being over-engineered perhaps. A little too generic naming throughout? Generic abstractions are sometimes either incredibly valuable or a signal to the contrary, I think. Especially when names like "compose" or "observer" or "refresh" or "transpose" (etc) pop up - they beg questioning 😁 

(not saying that the abstraction introduced was "bad", just that I found it a little harder to digest than a few conditions and some commenting around)

chriskmnds reviewed (commented):
Heya! Left a few more comments.

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,188 @@
+	const handleUpgrade = useUpgradeHandler( siteSlug, withDiscount, cartHandler );
+
+	return ( { planSlug, cartItemForPlan, selectedStorageAddOn }: UseActionCallbackParams ) => {
+		return ( isFreeTrialCta?: boolean ) => {
```
Do we need to add `isFreeTrialCta` as a parameter here? It looks like `planSlug` above should be enough to compare with `freeTrialPlanSlugs` and determine if it's a free trial action, no? 🤔

jeyip replied:
So there's a little bit of complexity here that I've gone back and forth about:

As it stands today, the `planSlug` you're referring to will always be the _non_ free trial plan slug. This is easily addressed when we call the `useActionCallback` hook by feeding it the `freeTrialPlanSlug` instead of the normal `planSlug`, but requires some updates that I wasn't entirely convinced about

I'll make the changes and link them here so we have something concrete to discuss.

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/64dfcd32144d74583e887c700d3d90e2902316d9

jeyip replied:
This change in particular is what I was hesitant about https://github.com/Automattic/wp-calypso/commit/64dfcd32144d74583e887c700d3d90e2902316d9#diff-00e19b2171f5efb1c76799b683dad83ab9c2512c3985ec84c20a7d62ad175dd8R467-R470. 

We initialize a call to `useActionCallback` with a free trial slug instead of a normal plan slug. It makes intuitive sense to me, but I had understood our vision of `useActionCallback` to centralize most if not all concerns about which cta function to invoke, and `isFreeTrialCta` was another one of those concerns. I'd love to hear your thoughts on this though

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -425,6 +364,40 @@ const PlansFeaturesMain = ( {
 
 	const eligibleForFreeHostingTrial = useSelector( isUserEligibleForFreeHostingTrial );
 
+	const planActionCallback = ( planSlug: PlanSlug ): boolean => {
```
I think let's just give it a weird, specific name for now and revisit later. There are a few options to move this out of here, but it may be something for a follow-up. It looks like we only need to move `isModalOpen` state  out of `plans-features-main` somehow.

I would name it something like `showModalAndExit` or `handleModalFlowAndReturn` or something/anything with  "modal" and "exit/return" in it. I guess anything that's not as generic as `planActionCallback`. WDYT?

jeyip replied:
Agree with your names. They're similar to the best I could come up with as well. https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1560341135

Also agree with the follow-up. I think it'll go by quickly when it's the subject of its own PR

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/1f9bd21389f71ac4a640a442bdd79b8382780616

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,188 @@
+			let upgradeCartItem = cartItemForPlan;
+
+			if ( isFreeTrialCta ) {
+				upgradePlanSlug = freeTrialPlanSlug as PlanSlug;
```
Is casting to `PlanSlug` here necessary? Can we avoid it?

jeyip replied:
If we proceed with https://github.com/Automattic/wp-calypso/pull/88694#discussion_r1562980127 this will no longer be necessary

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,188 @@
+				cartItemForPlan: upgradeCartItem,
+				isFreeTrialCta,
+				selectedStorageAddOn,
+			} );
```
NIT - I wonder if it'd be cleaner to use a switch for this last part instead, and call `handleUpgrade` in the `finally` clause?

jeyip replied:
I think so? The dedicated `default` clause helps clarify a bit.

I do wonder about having a series of `if/else` statements and suddenly changing the pattern to `switch/case`, but I feel like it's probably not a big deal, so let's try it.

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/a0893d47c8cbd4cbdc36f1527922160785f0cc55

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/a0893d47c8cbd4cbdc36f1527922160785f0cc55

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,188 @@
+import type { MinimalRequestCartProduct } from '@automattic/shopping-cart';
+
+function useUpgradeHandler(
+	siteSlug?: string | null,
+	withDiscount?: string,
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void
```
In general, I much prefer single object props (spread -> `{ prop1, prop2, etc. }`), rather than enumerating properties this way. It's easier to extend and maintain, order won't matter, and also more consistent with React environment.

jeyip replied:
Explanation makes sense, especially about consistency with react environment. Making the change 👍

jeyip replied:
[9989ba0](https://github.com/Automattic/wp-calypso/pull/88694/commits/9989ba040c427175bb136b39721e1acd998f0db6)

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,188 @@
+	);
+}
+
+function useGenerateActionCallback(
+	// TODO: Reevaluate param types and whether or not they should be optional
+	currentPlan: Plans.SitePlan | undefined,
+	eligibleForFreeHostingTrial: boolean,
+	cartHandler?: ( cartItems?: MinimalRequestCartProduct[] | null ) => void,
+	flowName?: string | null,
+	intent?: PlansIntent | null,
+	planActionCallback?: ( planSlug: PlanSlug ) => boolean,
+	sitePlanSlug?: PlanSlug | null,
+	siteSlug?: string | null,
+	withDiscount?: string
+): UseActionCallback {
```
Same. Can we convert this to an object prop (`{ currentPlan, cartHandler, etc. }`)?

jeyip replied:
Yeah we totally can -- I'm on it 👍

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/88694/commits/9989ba040c427175bb136b39721e1acd998f0db6

chriskmnds reviewed (commented):
Hey! Thanks for entertaining and addressing all the comments. I left a few more. It's looking good :-)

Happy to test with you more once we add the instructions. One thing I noticed is a disabled `manage plan` button in the spotlight plan when in `/plans/[paid site]`. Is this expected?

<img width="400" alt="Screenshot 2024-04-12 at 3 19 28 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/9c254b76-aecd-4fe4-8a39-6406ad827aac">

jeyip commented:
> Happy to test with you more once we add the instructions. One thing I noticed is a disabled manage plan button in the spotlight plan when in /plans/[paid site]. Is this expected?

Thanks! Since we've been making quite a few changes, I've only had time to do cursory testing between rounds of code review feedback.

I'll double check this bug soon

jeyip committed:
Use switch syntax for handleUpgrade condition block

jeyip committed:
Convert to single object params for useGenerateActionCallback hook

jeyip committed:
Rename `planActionCallback` to `showModalAndExit`

jeyip committed:
Remove `isFreeTrialCta` param

jeyip committed:
Fix types

jeyip committed:
Fix disabled manage plan button

jeyip committed:
Remove now unnecessary wrapping callback for Ctas

jeyip commented:
@chriskmnds the manage plan button bug should be fixed now. I wanted to note that the merge conflicts are becoming pretty unwieldy 😅 -- if you're busy Christos, no worries, but if you happen to have time, I'd be super appreciative if we could 
 get this PR tested together.

I didn't have time to test this PR thoroughly myself today, but I left some smoke testing notes in the PR description. I'll be taking a closer look tomorrow.

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-action-callback.ts:
```diff
@@ -0,0 +1,184 @@
+			}
+
+			/* 3. Handle plan upgrade and plan upgrade tracks events */
+			switch ( ! isFreePlan( planSlug ) ) {
+				case true:
+					recordTracksEvent?.( 'calypso_plan_features_upgrade_click', {
+						current_plan: sitePlanSlug,
+						upgrading_to: planSlug,
+						saw_free_trial_offer: !! freeTrialPlanSlug,
+					} );
+				case false:
+					recordTracksEvent( 'calypso_signup_free_plan_click' );
+				default:
+					handleUpgrade( {
+						cartItemForPlan,
+						selectedStorageAddOn,
+					} );
+			}
```
This will probably not work. I think it's a little trickier to achieve, so let's just go back to normal if/else :-)

chriskmnds commented:
> if you're busy Christos, no worries, but if you happen to have time, I'd be super appreciative if we could
> get this PR tested together.

Hey @jeyip ! Thanks for addressing the feedback. I think the switch statement is broken, can we maybe revert? The code looks good otherwise. 🙇 

I can help get this over the line, or unless someone else catches me to it. Thanks for the instructions!

jeyip committed:
Revert switch syntax in useActionCallback hook

jeyip committed:
Fix modal upsell cta

jeyip commented:
## Testing
### Requirements

For each of the flows listed below:
- Verified CTA behavior of the starter plan and creator plan if available
- Verified that the plan could be purchased and a new site could be created
- Verified behavior of the CTA in the features grid and comparison grid

Flows:
- [x] /start
- [x] /start/onboarding-pm
- [x] /setup/newsletter
- [x] /setup/new-hosted-site
- [x] /plans

### Browsers
- [x] Chrome

jeyip commented:
Testing looks good. Will merge this PR soon
